working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+expressive power, we get improved performance.
+
+Here is another example. It illustrates the fact that an extension type `E`
+with on-type `T` introduces a type `E` which is a supertype of `T`. This
+makes it possible to assign an expression of type `T` to a variable of type
+`E`. This corresponds to "entering" the extension type (accepting the
+specific discipline associated with `E`). Conversely, a cast from `E` to
+`T` is a downcast, and hence it must be written explicitly. This cast
+corresponds to "exiting" the extension type (allowing for violations of the
+discipline associated with `E`), and the fact that the cast must be written
+explicitly helps developers maintaining the discipline as intended, rather
+than dropping out of it by accident, silently.
+
+```dart
+extension type ListSize<X> on List<X> {
+  int get size => length;
+  X front() => this[0];
+}
+
+void main() {
+  ListSize<String> xs = <String>['Hello']; // OK, upcast.
+  print(xs); // OK, `toString()` available on Object?.
+  print("Size: ${xs.size}. Front: ${xs.front()}"); // OK.
+  xs[0]; // Error, `operator []` not a member of `ListSize`.
+
+  List<ListSize<String>> ys = [xs]; // OK.
+  List<List<String>> ys2 = ys; // Error, downcast.
+  ListSize<ListSize<Object>> ys3 = ys; // OK.
+  ys[0].front(); // OK.
+  ys3.front().front(); // OK.
+  ys as List<List<String>>; // `ys` is promoted, succeeds at run time.
+}
+```
+
+
+## Syntax
+
+The rule for `<extensionDeclaration>` in the grammar is replaced by the
+following:
+
+```ebnf
+<extensionDeclaration> ::=
+  'protected'? 'extension' ('type'? <typeIdentifier> <typeParameters>?)?
+      <extensionShowHidePart> 'on' <type> <interfaces>? '{'
+    (<metadata> <extensionMemberDefinition>)*
+  '}'
+
+<extensionShowHidePart> ::=
+  <extensionShowClause>? <extensionHideClause>?
+
+<extensionShowClause> ::= 'show' <extensionShowHideList>
+
+<extensionHideClause> ::= 'hide' <extensionShowHideList>
+
+<extensionShowHideList> ::=
+  <extensionShowHideElement> (',' <extensionShowHideElement>)*
+
+<extensionShowHideElement> ::=
+  <type> | <identifier> | 'operator' <operator>
+```
+
+*In the rule `<extensionShowHideElement>`, note that `<type>` derives
+`<typeIdentifier>`, which makes `<identifier>` nearly redundant. However,
+`<identifier>` is still needed because it includes some strings that cannot
+be the name of a type but can be the basename of a member, e.g., the
+built-in identifiers.*
+
+
+## Static Analysis
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type: It can occur as the
```

Comment:
What is the type of `this` inside of an extension type?  It feels like we will want it to have the extension type, not the `on` type, with free assignability between the extension type and the on type, but that doesn't match the existing semantics of extensions.  e.g.

```
  extension E on int {
     void test() {
         var x = this;  // What is the inferred type of `x`? I would expect `E`
         E y = 4; // I think this needs to work
         int z = this;  // I expect this needs to work inside of an extension as well
     }
}
```

Reply:
My starting point is that the type of `this` should be the on-type: The extension type provides some services based on invocations of instance members of the on-type, so that would be the main type of action.

So I'd expect the inferred type of `x` to be `int`. `E y = 4;` would be allowed (it is an upcast), and so would `int z = this;`.

With a protected extension type `E y = 4` would be a compile-time error (needs to call a constructor, so we also need to have some constructors).